src/test/java/org/zalando/nakadi/cache/SimpleCacheTest.java

```diff
@@ -0,0 +1,158 @@
+
+        // 2. In case of refresh, inform that item_1 has changed and item3 - deleted.
+        final Set<Value> expectedCallParams = new HashSet<>();
+        expectedCallParams.add(item1v1);
+        expectedCallParams.add(item2v1);
+        expectedCallParams.add(item3v1);
+
+        // Now inform that key1 changed and key3 was removed
+        Mockito.when(dataProvider.getFullChangeList(Mockito.eq(expectedCallParams)))
+                .thenReturn(new CacheChange(Arrays.asList("key1"), Arrays.asList("key3")));
+        cache.refresh();
+
+        // Now expect that key3 is resulting in null, key1 in second version, key2 - still first version
+        // (as it is not changed)
+        Assert.assertEquals(item1v2, cache.get("key1"));
+        Assert.assertEquals(item2v1, cache.get("key2"));
+        Assert.assertNull(cache.get("key3"));
+
+        Mockito.verify(dataProvider, Mockito.times(2)).load("key1");
+        Mockito.verify(dataProvider, Mockito.times(1)).load("key2");
+        Mockito.verify(dataProvider, Mockito.times(2)).load("key3");
+
+        // And now call again, to check again
+        Assert.assertEquals(item1v2, cache.get("key1"));
+        Assert.assertEquals(item2v1, cache.get("key2"));
+        Assert.assertNull(cache.get("key3"));
+
+        Mockito.verify(dataProvider, Mockito.times(2)).load("key1");
+        Mockito.verify(dataProvider, Mockito.times(1)).load("key2");
+        Mockito.verify(dataProvider, Mockito.times(3)).load("key3");
+    }
+
+    @Test
+    public void testInvalidationListeners() {
+        // There are 2 ways to get to invalidation listener - through invalidate call and through cache refresh.
+        // In order to achieve the result we will invalidate and record the invalidation listeners calls.
+
+        final CacheDataProvider dataProvider = Mockito.mock(CacheDataProvider.class);
+        final SimpleCache cache = new SimpleCache<>(dataProvider, t -> t);
+        final List<String> recordedChanges = new ArrayList<>();
+        cache.addInvalidationListener((Consumer<String>) recordedChanges::add);
+
+        Mockito.when(dataProvider.load("key1")).thenReturn(new Value("key1", "1"));
+        Mockito.when(dataProvider.load("key2")).thenReturn(new Value("key1", "1"));
+        Mockito.when(dataProvider.load("key3")).thenReturn(new Value("key1", "1"));
+
+        cache.get("key1");
+        cache.get("key2");
+        cache.get("key3");
+
+        Assert.assertEquals(Collections.emptyList(), recordedChanges);
+        cache.invalidate("key1");
+        Assert.assertEquals(Arrays.asList("key1"), recordedChanges);
+        cache.get("key1"); // Let's put it again to cache
+
+        Mockito.when(dataProvider.getFullChangeList(Mockito.anySet())).thenReturn(
+                new CacheChange(Arrays.asList("key3"), Arrays.asList("key1")));
+        cache.refresh();
+        Assert.assertEquals(Arrays.asList("key1", "key1", "key3"), recordedChanges);
+    }
+}
```

Comment:
![](https://media.giphy.com/media/l3UcqjMBeQzXaoqGI/giphy.gif)

Reply:
Sorry, I didn't get the joke (((

Reply:
It's not visible here, but if you see it on "Files Changed" tab - you will see that there is no endline at the end of the file :)